source/common/formatter/substitution_formatter.cc

```diff
@@ -54,5 +54,104 @@ const re2::RE2& SubstitutionFormatParser::commandWithArgsRegex() {
   // clang-format on
 }
 
+JsonFormatBuilder::FormatElements
+JsonFormatBuilder::fromStruct(const ProtobufWkt::Struct& struct_format) {
+  output_.clear();
+
+  formatValueToFormatElements(struct_format.fields());
+  output_.push_back(FormatElement{std::move(buffer_), false});
+  ASSERT(buffer_.empty(), "buffer_ should be consumed");
+
+  return std::move(output_);
+};
+
+void JsonFormatBuilder::formatValueToFormatElements(const ProtobufWkt::Value& value) {
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::KIND_NOT_SET:
+  case ProtobufWkt::Value::kNullValue:
+    serializer_.addNull();
+    break;
+  case ProtobufWkt::Value::kNumberValue:
+    if (keep_value_type_) {
+      serializer_.addNumber(value.number_value());
+    } else {
+      // TODO(wbpcode): may use Buffer::Util::serializeDouble for best performance.
+      // But this function only be executed when the configuration is loading. So,
+      // it's fine to use the fmt::to_string.
+      serializer_.addString(fmt::to_string(value.number_value()));
```

Comment:
any reason not to just be consistent?

Reply:
Also I'm pretty confused why we are using the serializer during config.

Reply:
> any reason not to just be consistent?

`serializeDouble` require a output parameter to write the data which means if we just want to serialize a single number to string, we need to create another wrapper to provide `add` method. It make things complex and make no big sense.

> Also I'm pretty confused why we are using the serializer during config.

I think I explain this question repeatedly, and https://github.com/envoyproxy/envoy/pull/35545#issuecomment-2272759820 has a clear presentation.

Specific to number, we needn't to do sanitizing, but for the input formatter configuration (Protobuf::Struct), we need do the sanitizing.

There core idea is very simple and clear:
1. serialize the struct to JSON string when loading configuration.
2. Split out part with substitution format commands when loading configuration.
3. Evaluates  the command in the 2 and joins different parts to final output.

To ensure we handle the substitution format commands correctly, we iterate the input `Protobuf::Struct` manually to the do the 1 and 2.

An exist case in current envoy code base the application log. It only do this coarsely but share same core idea with this PR. See the `Logger::Registry::setJsonLogFormat`.